src/webgpu/api/operation/rendering/3d_texture_slices.spec.ts

```diff
@@ -0,0 +1,363 @@
+            }
+            `,
+        }),
+        entryPoint: 'main',
+      },
+      fragment: {
+        module: this.device.createShaderModule({
+          code: `
+            struct Output {
+              ${locations}
+            }
+            @fragment fn main() -> Output {
+              var output : Output;
+              ${outputs}
+              return output;
+            }
+            `,
+        }),
+        entryPoint: 'main',
+        targets,
+      },
+      primitive: { topology: 'triangle-list' },
+    });
+  }
+
+  getBufferSizeAndOffset(
+    attachmentWidth: number,
+    attachmentHeight: number,
+    attachmentCount: number
+  ): { bufferSize: number; bufferOffset: number } {
+    const bufferSize =
+      (attachmentCount * attachmentHeight - 1) * kBytesPerRowAlignment + attachmentWidth * 4;
+    const bufferOffset = attachmentCount > 1 ? attachmentHeight * kBytesPerRowAlignment : 0;
+    return { bufferSize, bufferOffset };
+  }
+
+  checkAttachmentResult(
+    attachmentWidth: number,
+    attachmentHeight: number,
+    attachmentCount: number,
+    buffer: GPUBuffer
+  ) {
+    const { bufferSize, bufferOffset } = this.getBufferSizeAndOffset(
+      attachmentWidth,
+      attachmentHeight,
+      attachmentCount
+    );
+    const expectedData = new Uint8Array(bufferSize);
+    for (let i = 0; i < attachmentCount; i++) {
+      for (let j = 0; j < attachmentHeight; j++) {
+        for (let k = 0; k < attachmentWidth; k++) {
+          expectedData[i * bufferOffset + j * 256 + k * 4] = k < j ? 0x00 : 0xff;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 1] = k < j ? 0xff : 0x00;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 2] = 0x00;
+          expectedData[i * bufferOffset + j * 256 + k * 4 + 3] = 0xff;
+        }
+      }
+    }
+
+    this.expectGPUBufferValuesEqual(buffer, expectedData);
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('one_color_attachment,mip_levels')
+  .desc(
+    `
+  Render to a 3d texture slice with mip levels.
+  `
+  )
+  .params(u => u.combine('mipLevel', [0, 1, 2]).combine('depthSlice', [0, 1]))
+  .fn(t => {
+    const { mipLevel, depthSlice } = t.params;
+
+    const texture = t.create3DTexture({
+      width: kSize << mipLevel,
+      height: kSize << mipLevel,
+      depthOrArrayLayers: 2 << mipLevel,
+      mipLevelCount: mipLevel + 1,
+      usage: GPUTextureUsage.RENDER_ATTACHMENT | GPUTextureUsage.COPY_SRC,
+    });
+
+    const { bufferSize } = t.getBufferSizeAndOffset(kSize, kSize, 1);
+
+    const buffer = t.device.createBuffer({
+      size: bufferSize,
+      usage: GPUBufferUsage.COPY_SRC | GPUBufferUsage.COPY_DST,
+    });
+    const pipeline = t.createRenderPipeline();
+
+    const colorAttachment = t.getColorAttachment(texture, {
+      baseMipLevel: mipLevel,
+      mipLevelCount: 1,
+    });
+    colorAttachment.depthSlice = depthSlice;
+
+    const encoder = t.createEncoder('non-pass');
```

Comment:
nit: No need to use the createEncoder helper unless you're going to use any of its functionality. It's only there to abstract over multiple encoder types, and check different places errors can come from.